### PR TITLE
Add support for EDE caching in cachedb and subnetcache

### LINKED ISSUE #873: EDE information missing from reply to upstream
**Describe the bug**
Asked this question in the unbound mailing list (https://lists.nlnetlabs.nl/pipermail/unbound-users/2023-April/008069.html).
George (Yorgos) Thessalonikefs answered this is possible by using "EDE codes for DNSSEC validation failures to the SERVFAIL answers".

The FTL developer has added code to evaluate these codes, however, the results are inconsistent.

After investigating this, it turns out **the EDE codes are NOT always present in the reply**, the pi-hole topic has an attachment whith extended **log**ging from pihole-FTL (= dnsmasq) and a **pcap** file that clearly shows the missing EDE information. The pi-hole topic, start reading from here: https://discourse.pi-hole.net/t/dnssec-discussion-support-for-proxy-dnssec/62217/17

The **pcap** file and **log** is found in the next entry (18)

**To reproduce**
run unbound with "ede: yes", check the replies from unbound -> downstream.
- dig www.dnssec-failed.org (on linux host) -> OK
- browse (Microsoft Edge or Firefox) to www.dnssec-failed.org -> fail

**Expected behavior**
return EDE information for all queries, if available.
the strange thing is it works (EDE is returned) for dig requests (dig A www.dnssec-failed.org and dig AAAA www.dnssec-failed.org), but fails when a browser is making the request (open www.dnssec-failed.org in Microsoft edge and / or Firefox)

**System:**
 - Unbound version: 1.17.2 (compiled from GitHub source)
 - OS: raspbian on raspberry pi 3B (Linux raspberrypi 6.1.21-v7+ #1642 SMP Mon Apr  3 17:20:52 BST 2023 armv7l GNU/Linux)
 - `unbound -V` output:
```
Version 1.17.2
Configure line: --prefix=/usr --sysconfdir=/etc --disable-static --enable-tfo-client --enable-tfo-server --with-libevent --with-libhiredis --enable-cachedb --with-pidfile=/run/unbound.pid
Linked libs: libevent 2.1.12-stable (it uses epoll), OpenSSL 1.1.1n  15 Mar 2022
Linked modules: dns64 cachedb respip validator iterator
TCP Fastopen feature available
```
**Additional information**
The goal is to run pihole-FTL (= dnsmasq) with option proxy-dnssec. This appears to be working, but the EDE information isn't always present.

### PR DESCRIPTION
**Pull request by TCY16**
This PR is a child of #759.

### CODE REVIEW

File: cachedb/cachedb.c [hunk 1]
```diff
@@ -539,6 +557,27 @@ parse_data(struct module_qstate* qstate, struct sldns_buffer* buf)
...
+		if (ede->opt_len >= 2) {
+			qstate->return_msg->rep->reason_bogus =
+				sldns_read_uint16(ede->opt_data);
+		}
+
+		/* allocate space and store the error string and it's size */
+		if (ede->opt_len > 2) {
+			size_t ede_len = ede->opt_len - 2;
+
+			qstate->return_msg->rep->reason_bogus_str = regional_alloc(
```
Comment (wcawijngaards) [inline cachedb/cachedb.c:566]: The string is not zero terminated. If it should be zero terminated, it should have an extra +1 on the alloc size and a 0 written after the memcpy.

File: testdata/ede_caching_cachedb.tdir/ede_caching_cachedb.post [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# #-- ede_caching_cachedb.post --#
+# source the master var file when it's there
+[ -f ../.tpkg.var.master ] && source ../.tpkg.var.master
+# use .tpkg.var.test for in test variable passing
+[ -f .tpkg.var.test ] && source .tpkg.var.test
+
+# teardown
+. ../common.sh
+kill_pid $UNBOUND_PID
```
Comment (wcawijngaards) [inline testdata/ede_caching_cachedb.tdir/ede_caching_cachedb.post]: Also stop the second unbound instance. It was also started by the test.
```suggestion
kill_pid $UNBOUND_PID
kill_pid $UNBOUND_PID2
```

File: testdata/ede_caching_cachedb.tdir/ede_caching_cachedb.pre [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+. ../common.sh
+
+$PRE/unbound -V > cachedb_module_check.txt
+
+if ! grep -q -e "enable-cachedb" cachedb_module_check.txt
+then
+	echo "Missing cachedb_module in the ./configure stage:"
+	echo "'unbound -V' output:"
+	cat cachedb_module_check.txt
+	exit 1
```
Comment (wcawijngaards) [inline testdata/ede_caching_cachedb.tdir/ede_caching_cachedb.pre]: This function skips the test. It is called from the .pre file.
```suggestion
	skip_test "missing cachedb module in the ./configure stage"
```

File: testdata/ede_caching_cachedb.tdir/ede.conf [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+	port: @PORT@
+	use-syslog: no
+	directory: .
+	pidfile: "unbound.pid"
+	chroot: ""
+	username: ""
+	directory: ""
+	val-log-level: 2
+
+	auto-trust-anchor-file: "root.key"
```
Comment (wcawijngaards) [inline testdata/ede_caching_cachedb.tdir/ede.conf]: Is the trust anchor needed in this test?

File: testdata/ede_caching_subnetcache.tdir/ede.conf [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+	port: @PORT@
+	use-syslog: no
+	directory: .
+	pidfile: "unbound.pid"
+	chroot: ""
+	username: ""
+	directory: ""
+	val-log-level: 2
+
+	auto-trust-anchor-file: "root.key"
```
Comment (wcawijngaards) [inline testdata/ede_caching_subnetcache.tdir/ede.conf]: Is the trust anchor needed?

File: testdata/ede_caching_subnetcache.tdir/ede_caching_subnetcache.post [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# #-- ede_caching_subnetcache.post --#
+# source the master var file when it's there
+[ -f ../.tpkg.var.master ] && source ../.tpkg.var.master
+# use .tpkg.var.test for in test variable passing
+[ -f .tpkg.var.test ] && source .tpkg.var.test
+
+# teardown
+. ../common.sh
+kill_pid $UNBOUND_PID
```
Comment (wcawijngaards) [inline testdata/ede_caching_subnetcache.tdir/ede_caching_subnetcache.post]: The second unbound instance can be stopped here.
```suggestion
kill_pid $UNBOUND_PID
kill_pid $UNBOUND_PID2
```

File: testdata/ede_caching_subnetcache.tdir/ede_caching_subnetcache.pre [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+. ../common.sh
+
+$PRE/unbound -V > subnet_module_check.txt
+
+if ! grep -q -e "enable-subnet" subnet_module_check.txt
+then
+	echo "Missing subnet_module in the ./configure stage:"
+	echo "'unbound -V' output:"
+	cat subnet_module_check.txt
+	exit 1
```
Comment (wcawijngaards) [inline testdata/ede_caching_subnetcache.tdir/ede_caching_subnetcache.pre]: The skip_test function can be used here.
```suggestion
	skip_test "missing cachedb module in the ./configure stage"
```

File: iterator/iterator.c [hunk 8]
```diff
@@ -3809,6 +3810,9 @@ processFinished(struct module_qstate* qstate, struct iter_qstate* iq,
 	/* make sure QR flag is on */
 	iq->response->rep->flags |= BIT_QR;
 
+	/* explicitly set the EDE string size to 0 */
+	iq->response->rep->reason_bogus_str_size = 0;
```
Comment (wcawijngaards) [inline iterator/iterator.c]: It is more consistent to also update the reason_bogus_str to NULL.
```suggestion
	iq->response->rep->reason_bogus_str = NULL;
	iq->response->rep->reason_bogus_str_size = 0;
```

**Review by wcawijngaards — APPROVED** [review]
Looks good, added a couple suggestions that may make it nicer. It is nice to also support the cachedb and subnetcache for EDE error strings.

**Comment by jpgpi250** [comment]
@TCY16 

it has been a while since you updated this (and other pull requests). It has been suggested by @gthess in my issue as a possible solution for the inconsistencies I experience, ref issue #873 

I've compiled compiled the branche ede-caching-cachedb, there are some improvements, but I'm really waiting for all of your improvements to go into the master branch.

Could you please take a look at the suggested changes and conflicts, in order to get these valuable additions back on track?

Thanks for your time and effort.

File: testdata/ede_caching_cachedb.tdir/ede_caching_cachedb.post (hunk 2 above)
Comment (gthess) [reply testdata/ede_caching_cachedb.tdir/ede_caching_cachedb.post]: These tests no longer exist; mark as resolved.

File: testdata/ede_caching_cachedb.tdir/ede_caching_cachedb.pre (hunk 3 above)
Comment (gthess) [reply testdata/ede_caching_cachedb.tdir/ede_caching_cachedb.pre]: These tests no longer exist; mark as resolved.

File: testdata/ede_caching_subnetcache.tdir/ede.conf (hunk 5 above)
Comment (gthess) [reply testdata/ede_caching_subnetcache.tdir/ede.conf]: These tests no longer exist; mark as resolved.

File: testdata/ede_caching_subnetcache.tdir/ede_caching_subnetcache.post (hunk 6 above)
Comment (gthess) [reply testdata/ede_caching_subnetcache.tdir/ede_caching_subnetcache.post]: These tests no longer exist; mark as resolved.

File: testdata/ede_caching_subnetcache.tdir/ede_caching_subnetcache.pre (hunk 7 above)
Comment (gthess) [reply testdata/ede_caching_subnetcache.tdir/ede_caching_subnetcache.pre]: These tests no longer exist; mark as resolved.

File: testdata/ede_caching_cachedb.tdir/ede.conf (hunk 4 above)
Comment (gthess) [reply testdata/ede_caching_cachedb.tdir/ede.conf]: These tests no longer exist; mark as resolved.

File: iterator/iterator.c (hunk 8 above)
Comment (gthess) [reply iterator/iterator.c]: This is already resolved in #759.

File: cachedb/cachedb.c (hunk 1 above)
Comment (gthess) [reply cachedb/cachedb.c:566]: This is now resolved with https://github.com/NLnetLabs/unbound/pull/790/commits/c15cfb4bd933b993dbf2a6d3eaba4dba283e7e98

**Review by gthess — APPROVED** [review]
After addressing @wcawijngaards feedback and the recent changes, this seems ready now.

File: services/cache/dns.c [hunk 9]
```diff
@@ -654,6 +655,10 @@ tomsg(struct module_env* env, struct query_info* q, struct reply_info* r,
 	msg->rep->rrset_count = r->rrset_count;
 	msg->rep->authoritative = r->authoritative;
 	msg->rep->reason_bogus = r->reason_bogus;
+	if(r->reason_bogus_str) {
+		msg->rep->reason_bogus_str = regional_strdup(region, r->reason_bogus_str);
```
Comment (wcawijngaards) [inline services/cache/dns.c:659]: The variable `msg->rep->reason_bogus_str` was not zeroed by `gen_dns_msg`, so it should be set to NULL when `r->reason_bogus_str` is NULL, I think?

File: cachedb/cachedb.c [hunk 10]
```diff
@@ -545,6 +553,24 @@ parse_data(struct module_qstate* qstate, struct sldns_buffer* buf)
...
+			qstate->return_msg->rep->reason_bogus =
+				sldns_read_uint16(ede->opt_data);
+		}
+		/* allocate space and store the error string and it's size */
+		if(ede->opt_len > 2) {
+			size_t ede_len = ede->opt_len - 2;
+			qstate->return_msg->rep->reason_bogus_str = regional_alloc(
+				qstate->region, sizeof(char) * (ede_len+1));
+			memcpy(qstate->return_msg->rep->reason_bogus_str,
+			ede->opt_data+2, ede_len);
```
Comment (wcawijngaards) [inline cachedb/cachedb.c]: Perhaps this is better indented.
```suggestion
				ede->opt_data+2, ede_len);
```

**Review by wcawijngaards — APPROVED** [review]
The intervening commits are good fixes, in my opinion.

File: services/cache/dns.c (hunk 9 above)
Comment (gthess) [reply services/cache/dns.c:659]: It is set to NULL in `gen_dns_msg`.

Comment (wcawijngaards) [reply services/cache/dns.c:659]: That is correct. I was looking in the master branch, for that.

**Pull request merged**